File: src/lexer.ts
import { syntaxError } from "./errors";
import { LEGAL_EXTRA_CHARS, OPERATOR_MAP, TOKEN_MAP, Token, TokenName } from "./token";

export function tokenize(code: string, sourceFile: string = "unknown"): Token[] {
    let currentChar = 0;
    let currentColumn = 0;
    let currentLine = 0;
    let docstring = "";
    let tokens: Token[] = [];
    let text: string | undefined = "";

    let bracketStack: { type: number, index: number }[] = []

    let munchTrailing = () => {
        let munchedAnything = false;
        while (currentChar < code.length && code.charCodeAt(currentChar) <= 32) {
            munchedAnything = true;
            if (code[currentChar] == '\n') {
                currentColumn = 0;
                currentLine++;
            }
            else {
                currentColumn++;
            }
            currentChar++;
        }

        return munchedAnything;
    }

    let munchComment = (): string | undefined => {
        let munchedText = "";

        if (!(currentChar < code.length - 1 && code[currentChar] == '/' && code[currentChar + 1] == '*')) {
            return;
        }

        currentChar += 2

        while (currentChar < code.length - 1 /* Sufficient space for munching the cositos */ && !(code[currentChar + 1] == '/' && code[currentChar] == '*')) {
            munchedText += code[currentChar];

            if (code[currentChar] == '\n') {
                currentColumn = 0;
                currentLine++;
            }
            else {
                currentColumn++;
            }
            currentChar++;
        }

        if (!(currentChar < code.length - 1)) {
            // unclosed trash
            syntaxError(`Unclosed comment at end of source`, currentChar, currentLine, currentColumn, code ?? "", sourceFile);
        }

        currentChar += 2

        return munchedText;
    }

    let munchNumber = (): string | undefined => {
        let munchedText = "";

        let counter = 0;

        while (currentChar < code.length) {
            if ((code.charCodeAt(currentChar) >= 48 && code.charCodeAt(currentChar) <= 57) || (counter && ['.', 'x', 'o', 'b'].includes(code[currentChar]))) {
                munchedText += code[currentChar];
                currentChar++;
                counter++;
                currentColumn++;
            }
            else {
                break
            }
        }

        return munchedText;
    }

    let munchString = (): string | undefined => {
        let munchedText = "";

        if (!(currentChar < code.length && code[currentChar] == '"')) {
            return;
        }

        currentChar++;

        while (currentChar < code.length && !(code[currentChar] == '"')) {
            munchedText += code[currentChar];

            if (code[currentChar] == '\n') {
                currentColumn = 0;
                currentLine++;
            }
            else {
                currentColumn++;
            }
            currentChar++;
        }

        if (!(currentChar < code.length)) {
            // unclosed trash
            syntaxError(`Unclosed string at end of source`, currentChar, currentLine, currentColumn, code ?? "", sourceFile);
        }

        currentChar++

        return munchedText;
    }

    let munchWord = (): string | undefined => {
        let munchedText = "";

        let counter = 0;

        while (currentChar < code.length) {
            if ((code.charCodeAt(currentChar) >= 65 && code.charCodeAt(currentChar) <= 90) || (code.charCodeAt(currentChar) >= 97 && code.charCodeAt(currentChar) <= 122) || (counter && (code.charCodeAt(currentChar) >= 48 && code.charCodeAt(currentChar) <= 57)) || code.charAt(currentChar) == '_') {
                munchedText += code[currentChar];
                currentChar++;
                counter++;
                currentColumn++;
            }
            else {
                break
            }
        }

        return munchedText;
    }

    while (currentChar < code.length) {
        let lastPosition = currentChar;

        if (munchTrailing()) {
            // Do Nothing Xd
        }
        else if (text = munchComment()) {
            docstring = text;
            // Useful for docs or something of that wtfness
            // console.log(docstring);
        }
        else if (text = munchNumber()) {
            tokens.push({
                kind: 'number',
                text: text + "",
                line: currentLine,
                column: currentColumn,
                index: currentChar
            })
        }
        else if (text = munchString()) {
            tokens.push({
                kind: 'string',
                text: text + "",
                line: currentLine,
                column: currentColumn,
                index: currentChar
            })
        }
        else if (text = munchWord()) {
            let info: any = TOKEN_MAP[text];

            tokens.push({
                kind: info?.kind || 'identifier',
                text: text + "",
                code: info?.code ?? -1,
                line: currentLine,
                column: currentColumn,
                index: currentChar
            })
        }
        else {
            text = "";
            let info = undefined;
            let tempinfo = undefined;

            /**
             * The green pipe black chained muncher (GPBCM)
             */

            while (currentChar < code.length && LEGAL_EXTRA_CHARS.has(code[currentChar])) {
                if ( (tempinfo = OPERATOR_MAP[text + code[currentChar]]) )
                {
                    info = tempinfo;
                    text += code[currentChar]
                }
                else
                {
                    break;
                }
                currentChar++;
                currentColumn++;
            }
            
            if (text.length <= 0 || !info)
            {
                syntaxError(`Illegal character sequence`, currentChar, currentLine, currentColumn, code ?? "", sourceFile);
                return [];
            }

            if (info.paren)
            {
                if (info.opening)
                {
                    bracketStack.push({ type: info.code ?? -1, index: tokens.length })
                }
                else
                {
                    let lastBracket = bracketStack[ bracketStack.length - 1 ];
                    if (info.alternate != lastBracket?.type)
                    {
                        syntaxError(`Bracket closing mismatch`, currentChar, currentLine, currentColumn, code ?? "", sourceFile);
                    }
                    tokens[ lastBracket.index ].closingPair = tokens.length;
                    bracketStack.pop()
                }
            }

            tokens.push({
                kind: info.kind ?? 'none',
                text: text + "",
                code: info.code,
                paren: !!info.paren,
                line: currentLine,
                column: currentColumn,
                index: currentChar
            })
        }
    }

    // Implementación
    return tokens;
}

File: src/lexer.cpp
#include "lexer.hpp"

namespace YY
{
    namespace Lexer
    {
        inline bool isWhitespace(char code)
        {
            return code == ' ' || code == '\n' || code == '\t' || code == '\r';
        }

        inline bool isDigit(char code)
        {
            return code >= '0' && code <= '9';
        }

        inline bool isAlpha(char code)
        {
            return (code >= 'A' && code <= 'Z') || (code >= 'a' && code <= 'z');
        }

        inline bool isWord(char code)
        {
            return isAlpha(code) || isDigit(code) || code == '_';
        }

        inline bool isWordStart(char code)
        {
            return isAlpha(code) || code == '_';
        }

        void Lexer::trim()
        {
            if (!content)
            {
                return;
            }

            while (cursor < length && isWhitespace(content[cursor]))
            {
                if (content[cursor] == '\n')
                {
                    line++;
                    bol = ++cursor;
                }
                else
                {
                    cursor++;
                }
            }
        }

        Token::Token Lexer::next()
        {
            Token::Token token = Token::Token();

            trim();

            token.text = &content[length - 1];

            if (cursor >= length)
            {
                return token;
            }

            if (content[cursor] == '"')
            {
                token.text = &content[cursor];
                token.kind = Token::Kind::C_STRING;
                token.length = 1;
                cursor++;
                while (cursor < length && content[cursor] != '"')
                {
                    cursor++;
                    token.length++;
                }
                token.length++;
                cursor++;
                return token;
            }
            else if (cursor < length - 1 && content[cursor] == '/' && content[cursor + 1] == '*')
            {
                token.text = &content[cursor];
                token.kind = Token::Kind::C_COMMENT;
                token.length = 2;
                cursor += 2;
                while (cursor < length - 1 && !(content[cursor + 1] == '/' && content[cursor] == '*'))
                {
                    cursor++;
                    token.length++;
                }
                token.length += 2;
                cursor += 2;
                return token;
            }
            else if (isWordStart(content[cursor]))
            {
                token.text = &content[cursor];
                token.kind = Token::Kind::T_IDENTIFIER;
                while (cursor < length && isWord(content[cursor]))
                {
                    cursor++;
                    token.length++;
                }
                if (auto keywordToken = Token::keywordTokens.find(std::string(token.text, token.length)); keywordToken != Token::keywordTokens.end())
                {
                    token.kind = keywordToken->second;
                }
                return token;
            }
            else if (auto literalToken = Token::literalTokens.find(content[cursor]); literalToken != Token::literalTokens.end())
            {
                token.kind = literalToken->second;
                token.text = &content[cursor++];
                token.length = 1;
                return token;
            }
            else
            {
                token.text = &content[cursor];
                token.kind = Token::Kind::T_INVALID;

                /**
                 * The green pipe black chained muncher (GPBCM)
                 */

                while (cursor < length)
                {
                    if (auto tempinfo = Token::specialTokens.find(std::string(token.text, token.length + 1)); tempinfo != Token::specialTokens.end())
                    {
                        token.kind = tempinfo->second;
                        token.length++;
                    }
                    else
                    {
                        break;
                    }
                    cursor++;
                }

                if (token.kind != Token::Kind::T_INVALID)
                {
                    return token;
                }
            }

            token.kind = Token::Kind::T_INVALID;
            token.text = &content[cursor];
            token.length = 1;

            cursor++;

            return token;
        }
    }
}